docs/contributing/making-a-pull-request.md

```diff
@@ -0,0 +1,450 @@
+established style used in the project. Doing so increases readability and
+also helps with making diffs easier to read for those who will review the pull
+request. Avoid making any large-scale style changes such as asking your IDE
+to re-format all code.
+
+Study the code that you are modifying well to ensure that you fully understand
+how it works before you try to change it. This will not only help you solve the
+problem you are trying to address but also minimize the risks of creating
+unintended side effects.
+
+!!! tip "Linters"
+
+    We are not currently using linters as part of the build process. However,
+    this does not mean that you cannot use a linter in your development
+    environment.  In fact, doing so is probably a good way to get close to the
+    code style used in Material for MkDocs. A linter does not replace careful
+    crafting of code, though.
+
+### Committing to a branch
+
+Development for pull requests is best done in a topic branch separate from the
+`master` branch. Create a new local branch with `git switch -c <name>` and
+commit your changes to this branch.
+
+When you want to push commits to your fork, you can do so with
+`git push -u origin <name>`. The `-u` argument is the short version of
+`--set-upstream`, which makes the newly created branch 'track' the branch with
+the same `<name>` in your fork. This means that then `pull` and `push` commands
+will work against that branch in your fork by default.
+
+### Merging concurrent changes
+
+If the work you do takes some time then the chances increase that changes will
+be made to the main repository while you work.It is probably a good idea to set
+up the original Material for MkDocs repository as an `upstream` repository for
+your local clone.
+
+This is what it might look like:
+
+```bash hl_lines="4"
+$ git remote -v
+origin	git@github.com:<your_username>/mkdocs-material-fork.git (fetch)
+origin	git@github.com:<your_username>/mkdocs-material-fork.git (push)
+$ git remote add upstream https://github.com/squidfunk/mkdocs-material.git
+$ git remote -v
+origin	git@github.com:alexvoss/mkdocs-material-fork.git (fetch)
+origin	git@github.com:alexvoss/mkdocs-material-fork.git (push)
+upstream	https://github.com/squidfunk/mkdocs-material.git (fetch)
+upstream	https://github.com/squidfunk/mkdocs-material.git (push)
+```
+
+After you have done this, you can pull any concurrent changes from the upstream
+repository directly into your clone and do any necessary merges there, then push
+them up to your fork. You will need to be explicit about which remote repository
+you want to use when you are doing a `pull`:
+
+```bash
+# making and committing some local changes
+push pull upstream master
+```
+
+This fetches changes from the `master` branch into your topic branch and merges
+them.
+
+!!! question "Rebase, merge, squash?"
+
+    @squidfunk, this does raise the question what strategy to advise people to
```

Comment:
This does raise the question we have had a few times about committing to `material/`. I will try to write something here so that I get to know the build process better. Ultimately, you may need to help me but let me get some words down on this.

Reply:
Many PRs don't include the build output and I've kind of given up on requiring it, because most users don't want to go through the build process. Thus, I just went ahead merged it and then committed the built file in `material`  again. It is ugly and definitely not the best possible way, but it was pragmatic. Let's see if more users go through the troubles of doing that, because we're essentially up'ing the requirements for PRs.

Reply:
I do not see how we can produce a text at the moment without making it much more complicated. Deleted that section for now. The question whether to commit `material/` has come up a couple of time and I was hoping we could have a piece to point to in the future but conflicting changes are so difficult to deal with, never mind describe in the abstract, that I do not much hope that we can illuminate matters here. It might be that in the end we need to point to books others have written? "Pro Git" has been my go-to book but I cannot claim that I have read it cover-to-cover. That is still on my guilt-pile.

Reply:
We mention that the actual theme resides in the folder [here](https://squidfunk.github.io/mkdocs-material/customization/?h=material#building-the-theme), but that might be too out-of-context when crafting a PR. As said, I think this PR is good to be published for now, it is already a tremendous help. At least IMHO.

Reply:
Marked it as ready and will request a review - which you can skip if you are happy. I am saving this discussion and turning it into a todo list as we are leaving some things open. One thing I will come to eventually is the whole logic of building the theme and populating the `material/` folder. I will have a look at Pylint again now because it seems this could be done fairly quickly.